### PR TITLE
Dma/4976 watcher eth support

### PR DESCRIPTION
**Pull request by tt-dma**

### CODE REVIEW

File: tt_metal/hw/firmware/src/erisc.cc [hunk 1]
```diff
@@ -89,12 +89,14 @@ void __attribute__((section("erisc_l1_code"))) ApplicationHandler(void) {
     for (uint32_t n = 0; n < NUM_NOCS; n++) {
         noc_local_state_init(n);
     }
+    DEBUG_STATUS('I');
```
Comment (tt-dma) [inline tt_metal/hw/firmware/src/erisc.cc:92]: Just added a couple sample DEBUG_STATUSes here, we can modify/expand these as we see what parts are important for debugging moving forward.

File: tt_metal/hw/inc/risc_common.h [hunk 2]
```diff
@@ -211,4 +211,25 @@ inline void breakpoint_(uint32_t line) {
...
 
+// Helper function to wait for a specified number of cycles, safe to call in erisc kernels.
+#if defined(COMPILE_FOR_ERISC)
+// Forward declaration to avoid circular dependency
+namespace internal_ {
+void __attribute__((section("code_l1"))) risc_context_switch();
+void disable_erisc_app();
+}
+#endif
+inline void riscv_wait(uint32_t cycles) {
```
Comment (tt-dma) [inline tt_metal/hw/inc/risc_common.h:222]: This is a modified version of the wait() function from ckernels.h, which was moved into the llk repo.

File: tests/tt_metal/tt_metal/unit_tests_common/watcher/test_noc_sanitize.cpp [hunk 3]
```diff
@@ -122,7 +164,16 @@ TEST_F(WatcherFixture, TestWatcherSanitize) {
...
-        this->RunTestOnDevice(RunTest, device);
-    }
+
+    // Only run on device 0 because this test takes down the watcher server.
+    this->RunTestOnDevice(RunTest, this->devices_[0]);
+}
+
+TEST_F(WatcherFixture, TestWatcherSanitizeEth) {
+    // Skip this test for now until we add a way to recover from the hang.
+    log_info(LogTest, "Skip this test until we can recover from the hang on an eth core.");
```
Comment (tt-dma) [inline tests/tt_metal/tt_metal/unit_tests_common/watcher/test_noc_sanitize.cpp:174]: This is issue will be a bit more involved to fix, filed a separate issue here: #5397 

We'll put that fix in as a separate PR, so just disable for now.

File: tests/tt_metal/tt_metal/unit_tests_common/watcher/test_waypoint.cpp [hunk 4]
```diff
@@ -47,6 +52,27 @@ static void RunTest(WatcherFixture* fixture, Device* device) {
         CoreRange(xy_start, xy_end),
         ComputeConfig{});
 
+    // Also run on ethernet cores if they're present
+    bool has_eth_cores = !device->get_active_ethernet_cores().empty();
+    // TODO: Change back to a single erisc_kid once the bug with CoreRange fast dispatch is fixed.
```
Comment (tt-dma) [inline tests/tt_metal/tt_metal/unit_tests_common/watcher/test_waypoint.cpp:57]: Separate issue: #5398

**Review by TT-billteng — APPROVED** [review]

File: tt_metal/llrt/watcher.cpp [hunk 5]
```diff
@@ -420,26 +443,40 @@ static void dump_core(FILE *f, std::map<int, bool>& used_kernel_names, WatcherDe
     fprintf(f, "Device %i, ", wdev->device_id_);
     fprintf(f, "Core %s:%s  ", core.str().c_str(), pad.c_str());
 
+    // Ethernet cores have a different mailbox base addr
+    bool is_eth_core = tt::llrt::is_ethernet_core(core, wdev->device_id_);
+    uint64_t mailbox_addr = MEM_MAILBOX_BASE;
+    if (is_eth_core)
+        mailbox_addr = eth_l1_mem::address_map::ERISC_MEM_MAILBOX_BASE;
```
Comment (pgkeller) [inline tt_metal/llrt/watcher.cpp:450]: use {} or keep on one line

File: tests/tt_metal/tt_metal/test_kernels/misc/watcher_waypoints.cpp [hunk 6]
```diff
@@ -3,16 +3,18 @@
...
 
 #include <cstdint>
-#include "ckernel.h"
 #include "debug/status.h"
 
 // Helper function to sync execution by forcing other riscs to wait for brisc, which in turn waits
 // for a set number of cycles.
 void hacky_sync(uint32_t sync_num, uint32_t wait_cycles, uint32_t sync_addr) {
     volatile tt_l1_ptr uint32_t* sync_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(sync_addr);
+#if defined(COMPILE_FOR_BRISC) | defined(COMPILE_FOR_ERISC)
```
Comment (pgkeller) [inline tests/tt_metal/tt_metal/test_kernels/misc/watcher_waypoints.cpp:12]: || ?

File: tt_metal/hw/firmware/src/erisc.cc [hunk 7]
```diff
@@ -114,6 +116,7 @@ void __attribute__((section("erisc_l1_code"))) ApplicationHandler(void) {
         if (erisc_info->launch_user_kernel == 1) {
             kernel_profiler::init_profiler();
             kernel_profiler::mark_time(CC_MAIN_START);
+            DEBUG_STATUS('K');
```
Comment (pgkeller) [inline tt_metal/hw/firmware/src/erisc.cc:119]: take a look in, eg, brisc.cc and make these statuses match those, eg, 'R' for run and 'D' for done

File: tt_metal/hw/inc/debug/sanitize_noc.h [hunk 8]
```diff
@@ -30,6 +30,13 @@ enum debug_sanitize_which_riscv {
 
 #if defined(WATCHER_ENABLED)
 
+#if defined(COMPILE_FOR_ERISC)
+// Forward declaration to avoid circular dependency
+namespace internal_ {
+void __attribute__((section("code_l1"))) risc_context_switch();
```
Comment (pgkeller) [inline tt_metal/hw/inc/debug/sanitize_noc.h:36]: isn't this declared elsewhere?  why do we need to re-declare it here?  I see the forward declaration comment, but isn't there another header w/ it?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 9]
```diff
@@ -1016,7 +1016,7 @@ void HWCommandQueue::finish() {
             } else if (tt::llrt::watcher_server_killed_due_to_error()) {
                 this->exit_condition = true;
                 TT_THROW(
-                    "Command Queue could not finish: device hang due to illegal NoC transaction. See build/watcher.log "
+                    "Command Queue could not finish: device hang due to illegal NoC transaction. See built/watcher.log "
```
Comment (pgkeller) [inline tt_metal/impl/dispatch/command_queue.cpp:1019]: haha, nice

Comment (pgkeller) [reply tt_metal/impl/dispatch/command_queue.cpp:1019]: lets define the path in watcher.h in case we move it again...

File: tt_metal/hw/firmware/src/erisc.cc (hunk 7 above)
Comment (tt-dma) [reply tt_metal/hw/firmware/src/erisc.cc]: Updated these messages in latest commit

File: tt_metal/llrt/watcher.cpp (hunk 5 above)
Comment (tt-dma) [reply tt_metal/llrt/watcher.cpp:451]: Fixed

File: tt_metal/hw/inc/debug/sanitize_noc.h (hunk 8 above)
Comment (tt-dma) [reply tt_metal/hw/inc/debug/sanitize_noc.h:36]: This is to avoid an annoying circular dependency, this file includes `ethernet/dataflow_api.h`, but to use the sanitization functions there `ethernet_dataflow_api.h` needs to include this function. Perhaps going forward we could look to restructure some files to avoid having to do this

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 9 above)
Comment (tt-dma) [reply tt_metal/impl/dispatch/command_queue.cpp]: Fixed!

File: tt_metal/hw/inc/debug/sanitize_noc.h (hunk 8 above)
Comment (tt-dma) [reply tt_metal/hw/inc/debug/sanitize_noc.h:36]: Filed issue #5452 for this so we don't forget

**Comment by tt-dma** [comment]
Ci passing: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7939207011

**Comment by tt-dma** [comment]
CI Passes, I think it's better to get this PR in and handle any newly filed issues in a separate PR
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8024135283

**Review by pgkeller — APPROVED** [review]

File: tt_metal/hw/inc/ethernet/erisc.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
```
Comment (aliuTT) [inline tt_metal/hw/inc/ethernet/erisc.h:2]: I may change this to a different file name. just an fyi, no action needed

File: tt_metal/hw/inc/risc_common.h [hunk 11]
```diff
@@ -211,4 +211,21 @@ inline void breakpoint_(uint32_t line) {
...
+#include "erisc.h"
+#endif
+inline void riscv_wait(uint32_t cycles) {
+    volatile uint tt_reg_ptr * clock_lo = reinterpret_cast<volatile uint tt_reg_ptr * >(RISCV_DEBUG_REG_WALL_CLOCK_L);
+    volatile uint tt_reg_ptr * clock_hi = reinterpret_cast<volatile uint tt_reg_ptr * >(RISCV_DEBUG_REG_WALL_CLOCK_H);
+    uint64_t wall_clock_timestamp = clock_lo[0] | ((uint64_t)clock_hi[0]<<32);
+    uint64_t wall_clock = 0;
+        do {
+#if defined(COMPILE_FOR_ERISC)
+            internal_::risc_context_switch();
```
Comment (aliuTT) [inline tt_metal/hw/inc/risc_common.h:225]: are we letting users reset machine when this case is hit? Is so, make sure it's in the docs, https://tenstorrent-metal.github.io/tt-metal/latest/tools/watcher.html#watcher

File: tt_metal/llrt/watcher.cpp [hunk 12]
```diff
@@ -620,6 +671,23 @@ void watcher_init(int device_id,
         }
     }
 
+    // Initialize ethernet cores debug values
+    for (const CoreCoord &eth_core : ethernet_cores()) {
```
Comment (aliuTT) [inline tt_metal/llrt/watcher.cpp:675]: Is this looping over all ethernet cores? I think it is ok for now but Umair is putting some custom dispatch kernels to inactive cores. Please sync with him to make sure he's aware that the eth mailbox address is going to be written to.

**Review by aliuTT — APPROVED** [review]

File: tt_metal/hw/inc/risc_common.h (hunk 11 above)
Comment (tt-dma) [reply tt_metal/hw/inc/risc_common.h:225]: Not exactly sure what you mean here? This function is just a kernel wait function, I don't think we expect user's to call this. In terms of resetting on watcher sanitization hang, I think the idea is that I'll work on #5397 right away so users won't have to reset when that's hit. Let me know if I missed the meaning here!

File: tt_metal/llrt/watcher.cpp (hunk 12 above)
Comment (tt-dma) [reply tt_metal/llrt/watcher.cpp:675]: Will do!

Comment (tt-dma) [reply tt_metal/llrt/watcher.cpp:680]: Correction: this is active ethernet cores only - watcher will only poll active ethernet cores
